### ISSUE TITLE
Getting an entire grammar as a string

### ISSUE THREAD

**Issue by TMiguelT** [pyparsing/pyparsing#211]
I want to output my entire grammar in a vaguely human-readable format, so people can see what my parser is doing without having to necessarily understand pyparsing.

If I `print(str(grammar))` I get:
```
{{{LineStart Suppress:(<SP><TAB><CR><LF>)} flag_block} ^ {Suppress:(":") flag_block}}
```
This is a good start, but it's not recursively expanding tokens like `flag_block`. Is there any way to force this to happen in order to get a complete description of the grammar?

**Comment by TMiguelT** [comment]
I think this is happening because I've explicitly `setName()` on `flag_block`. Is there a way to ignore the token name when converting to string?

**Comment by ptmcg** [comment]
No, the purpose of setName is exactly to work as you describe. It helps make an expression like `Word(alphas, alphanums).setName("identifier")` to show up in exceptions as "Expected identifier" instead of "Expected W:(ABCD....)".  There have been other suggestions in the past also about displaying pyparsing parsers using "railroad track" diagrams, but even that effort would eventually look to some mechanism like setName that would support the definition of a friendly name for an expression, instead of a cryptic shorthand. Which items should be shown as names vs. their parsing constructs? Ultimately, you control that by which expressions you name using setName.

**Comment by TMiguelT** [comment]
Well since I want the full grammar, I was hoping that every "composite" token (and, or, sequence) could be decomposed to their contents, but atomic components (literal, regex, anything that implements its own parse function) should stay as they are since their internal behaviour can't really be represented in a grammar. 

A railroad diagram would be super useful as well.

I guess from your answer, there is no mechanism to do this? What I might do is write a function that descends the grammar tree, runs `delattr(token, 'name')`, and then calls `str()` on the entire grammar which should mean that all tokens are decomposed wherever possible. Would this work?

**Comment by ptmcg** [comment]
> What I might do is write a function that descends the grammar tree, runs delattr(token, 'name'), and then calls str() on the entire grammar which should mean that all tokens are decomposed wherever possible. Would this work?

Yes, I think that would work. You will get a long ugly mess though.

**Comment by TMiguelT** [comment]
Hmm well railroad diagrams certainly would be ideal. If I wrote a pyparsing railroad generator for them using something [this library](https://github.com/tabatkins/railroad-diagrams/blob/gh-pages/README-py.md), do you think that should be something I PR into the core? Maybe behind an `extras` install such that you `pip install pyparsing[railroad]`?

**Comment by ptmcg** [comment]
Yes, I think an `extras` install sounds appropriate (and I'm interested to see how this packaging is done). Right now I have a `recurse()` method in development for navigating the structure of a grammar without having to figure out if it should be `self.expr` or `self.exprs` or nothing to get to subcomponents of the grammar. Any element supports `recurse()` to return a list of its subelements. (`recurse` may not be the best name - `subelements` might be better. Would prefer not `children`.) In any event, I think a common method on all elements to get its sub-elements, if any, would be useful in support of a railroad diagrammer. Need to make sure we don't recurse forever with `Forward`s, but otherwise, should not be difficult.

**Comment by TMiguelT** [comment]
So do you think I should wait until you have `recurse()` fully implemented before I try generating diagrams?

**Comment by ptmcg** [comment]
You could get started by monkeypatching a method onto ParserElement that does something like:

    if hasattr(self, 'expr') and self.expr is not None:
        yield self.expr
    elif hasattr(self, 'exprs') and self.exprs is not None:
        yield from self.exprs
    else:
        pass

It will fail miserably with infinite recursion on `Forward`s, but would let you make some beginning progress. (How *do* you represent recursion in a railroad diagram, anyway?)

**Comment by TMiguelT** [comment]
Okay that sounds good. Why would Forwards cause infinite recursion? Won't `exprs` just return whatever was `<<` shifted into it? If recursion is intentional, possibly it could be represented as a loop/repeat instead?

**Comment by ptmcg** [comment]
The whole point of `Forward`'s is that somewhere down in that structure that get's `<<`'ed, there is another reference to that same `Forward`. That's how you get a recursive grammar to work. `nestedExpr` generates a simple recursive expression, `infixNotation` a more complicated one. Or grep for `Forward` in the examples directory, I'm sure some other cases will pop up.

**Comment by ptmcg** [comment]
`countedArray` uses a `Forward` also, but not for recursion, but for dynamically changing the grammar at parse time.

**Comment by TMiguelT** [comment]
I think if you have a recursive grammar you have to use the nonterminal's name in the diagram itself. For example: 
![image](https://user-images.githubusercontent.com/5019367/81898229-9e192780-95fb-11ea-9994-6f95013cf0cd.png)

**Comment by TMiguelT** [comment]
Closed by #218. Although it doesn't technically answer the original question, but it answers a better question!
